webui/react/src/components/BreadcrumbBar.tsx

```diff
@@ -0,0 +1,210 @@
+import { Breadcrumb, Tooltip } from 'antd';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import usePolling from 'hooks/usePolling';
+import { paths } from 'routes/utils';
+import { getExperimentDetails, getProject, getTrialDetails, getWorkspace } from 'services/api';
+import Icon from 'shared/components/Icon/Icon';
+import { isEqual } from 'shared/utils/data';
+import { ExperimentBase, Project, TrialDetails, Workspace } from 'types';
+import handleError from 'utils/error';
+
+import css from './BreadcrumbBar.module.scss';
+import Link from './Link';
+import WorkspaceIcon from './WorkspaceIcon';
+
+interface Props {
+  experiment?: ExperimentBase;
+  extra?: React.ReactNode;
+  id: number;
+  project?: Project;
+  trial?: TrialDetails;
+  type: 'project' | 'experiment' | 'trial';
+  workspace?: Workspace;
+}
```

Comment:
**question:** So this is a "smart" component where it's doing a lot of API calls given types and entities and then figuring out how to render after collecting all the info. Curious if there is a strong reason to approach it this way. One possibility is that we want to centralize the logic of fetching data for filling out the breadcrumb component.

Alternatives could be to have a context that keeps track of traversed entities, and at each traversal into specifics, to save the context of that particular item. (possibly in `Store` as something like `breadcrumbs`). For example, when traversing into an experiment detail, `breadcrumbs` would be `[ <experiment> ]`. Then diving into one of the trials, would be something like `breadcrumbs = [ <experiment x>, <trial x> ]`, etc. BreadcrumbBar can just consume the store directly, with the goal of avoid having to (potentially extra) API calls.

Reply:
Yeah, I wanted to have all of the logic for constructing the breadcrumbs in one place, and not have to worry about different pages handling it differently. With putting data into the Store, I worry about that somehow getting into a bad state and showing incorrect/stale data. Open to hearing other ideas though.